language-plutus-core/src/Language/PlutusCore/Quote.hs

```diff
@@ -0,0 +1,49 @@
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module Language.PlutusCore.Quote (
+              runQT
+            , runQ
+            , freshName
+            , freshTyName
+            , parse
+            , QT
+            , Q
+            ) where
+
+import           Control.Monad.Except
+import           Control.Monad.State
```

Comment:
The strict state monad can cause some problems with recursion, so I avoid it unless I have benchmarks that show a substantial improvement.

Reply:
After I fixed like ten space leaks, I started to believe that laziness should be explicit. My bet is that with the current approach: lazy `ByteString`s (that leak [literally always](https://ghc.haskell.org/trac/ghc/ticket/13630)) for names instead of `ShortByteString`s (again, see the link, it's a terrible scenario)), standard Haskell tuples as states instead of strict tuples or strict data types, the `State` itself being lazy, lazy data structures, etc, etc, we will eventually run into situation where some user complains that an attempt to type check his not that big program results in OOM.

Reply:
We use `Unique` for names.

Currently if you replace the lazy state monad with a strict one in `Language.PlutusCore.TypeSynthesis` it hangs indefinitely.

Reply:
> We use Unique for names.

We store `ByteString`s as well as `Unique`s.

> Currently if you replace the lazy state monad with a strict one in Language.PlutusCore.TypeSynthesis it hangs indefinitely.

Due to `repeatM`?

Reply:
I do not understand why `Name` is going to cause space leaks. Do you have a reason to believe this is occurring?

Reply:
I have no strong feelings - but we can change this afterwards easily.

Reply:
@vmchale Looking at the link, it seems like lazy bytestrings don't get GCd very well. That does seem like it would be a potential space leak.

Reply:
@vmchale, check the link:

> So it seems that I can get into the unfortunate situation where a super short ByteString of a few bytes can waste an entire 4 KB block of memory; some migth call this a leak.

Since we store `ByteString`s in `IntMap`s, that's a source for potential space leaks. Lazy `ByteString` are for streaming, really, they shouldn't be used to store short names. Check [the docs](https://hackage.haskell.org/package/bytestring-0.10.8.2/docs/Data-ByteString-Lazy.html):

> A key feature of lazy ByteStrings is the means to manipulate large or unbounded streams of data without requiring the entire sequence to be resident in memory. To take advantage of this you have to write your functions in a lazy streaming style, e.g. classic pipeline composition. The default I/O chunk size is 32k, which should be good in most circumstances.

I would use [`ShortText`](https://hackage.haskell.org/package/text-short-0.1.1/docs/Data-Text-Short.html) (which is defined in terms of `ShortByteString`) where we use lazy `ByteString`s currently. [Related blog post](https://markkarpov.com/post/short-bs-and-text.html).